### DISCUSSION TITLE
Export-Templates containing more than one object type

### DISCUSSION THREAD

**Discussion by reduzent** [netbox-community/netbox#13485] [Help Wanted!] [answered]
I would like to create an export template for DHCP reservations. However, there are two kind of of reservations, one is IP address based, the other hostname based. This means I need to access IPAddress objects for the former and Interface objects for the latter. Is that possible at all?

The editing page for export templates allows to define several "Content types", but obviously the `queryset` variable contains only one. What is the ability to set more than one "Content type" useful for, then?

**Comment by mtinberg** [comment]
IPAddress records come with a dns_name field that you can fill out, by hand or by writing custom code, they also have assigned_object (at least in the REST API) which includes info about the device/interface or FHRP group they are assigned to.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Roman Haefeli ***@***.***>
Sent: Wednesday, August 16, 2023 10:13 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Eport-Templates containting more than one Object type (Discussion #13485)


I would like to create an export template for DHCP reservations. However, there are two kind of of reservations, once is IP address based, the other hostname based. This means I need to access IPAddress objects for the former and Interface objects for the latter. Is that possible at all?

The editing page for export templates allows to define several "Content types", but obviously the queryset variable contains only one. What is the ability to set more than one "Content type" useful for, then?

**Reply by reduzent** [reply]
Thanks for your answer. We cannot use the IPAddress object for hostname based reservations because the DHCP then assigns an available IP address that netbox doesn't know about. 

I still wonder about why it is possible to set more than one content type for export templates.

**Comment by mtinberg** [comment]
I don't know what DHCP server software you are talking about, but it sounds like there is some misunderstanding about what you are trying to configure.  DHCP statics map an interface MAC address to an IP Address, so I'm not sure where DNS names come in except as a way to correlate config between the DNS service and the DHCP service, like an Infoblox Host record.

As for as why allowing more than one content type for export templates, why not?  If you either are exporting fields that are common to a number of data models (like name, comments, slug, id, etc) you might have a template that works for multiple things, or you might use if/then logic in the template to make it work for different data models.  I'm not sure there is a benefit to Netbox restricting this kind of flexability.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by reduzent** [reply]
> As for as why allowing more than one content type for export templates, why not?  If you either are exporting fields that are common to a number of data models (like name, comments, slug, id, etc) you might have a template that works for multiple things, or you might use if/then logic in the template to make it work for different data models.  I'm not sure there is a benefit to Netbox restricting this kind of flexability

I'm not all against many content types in export templates. I just haven't figured out, how to to use them within the export template itself. The `queryset` variable seems to hold only a list of models of the first specified type.

If I  - for instance - specify "IPAM -> IPAdress" , "DCIM -> Interface", how can I iterate through all interfaces?

**Comment by bitcollector1** [comment]
If you are using KEA DHCP Server, you can just make the reservation right from NetBox GUI using the device interface info

```
import requests
from dcim.models import Device, Interface
from ipam.models import Prefix
from extras.scripts import *

class ReservationUpdateScript(Script):

    class Meta:
        name = "KEA Reservation Update."
        description = "Update KEA Reservation"

    device = ObjectVar(
        description="host to check reservation",
        model=Device
    )

    interface = ObjectVar(
        description="(Required) Interface for Reservation",
        model=Interface,
        query_params={'device_id': '$device'}
    )

    def run(self, data, commit):
        
        device = data['device']
        interface = data['interface']
        
        interface = Interface.objects.get(name=interface.name, device__name=device.name)
        interface_ip = interface.ip_addresses.first()

        if 'bmc' in interface.name:
            hostname = f"{device.serial}-bmc"
        else:
            hostname = f"{device.serial}-{device.cf['sku']}"


        if interface_ip:
            subnet = Prefix.objects.get(prefix__net_contains_or_equals=str(interface_ip.address.ip), _depth=1)

            s = requests.Session()
            s.verify = '/etc/ssl/certs'

            headers = {
                'Content-Type': 'application/json',
            }

            json_data = {
                'command': 'reservation-update',
                "arguments": {
                    "reservation": {
                    'subnet-id': subnet.id,
                    'hw-address': str(interface.mac_address),
                    'ip-address': str(interface_ip.address.ip),
                    'hostname': hostname
                    }
                },
                'service': [
                    'dhcp4',
                ],
            }

            response = s.post('https://kea.vip.me.com/', headers=headers, json=json_data, auth=('user', 'password'))

            test = response.json()[0]['text']

            self.log_success(test)

        else:
            self.log_warning("No IP Found on Interface")
```

**Reply by reduzent** [reply]
Thanks, that's an interesting snippet. We're indeed using ISC Kea DHCP.

**Comment by bitcollector1** [comment]
<img width="1026" alt="Screenshot 2023-08-16 at 23 09 36" src="https://github.com/netbox-community/netbox/assets/50723251/e58f4c3e-3b34-4fce-8133-93f8257215c6">

**Comment by mtinberg** [comment, marked as answer]
The queryset is for the page you are looking at, so it is only ever returning one object type at a time, so if you browse /dcim/interfaces?device_id=123&enabled=True then the queryset matches what you've requested. Very similar to issuing a query using nbshell

>>> Interface.objects.filter(device_id=123, enabled=True)
<RestrictedQuerySet [<Interface: Ethernet1/1>, <Interface: Ethernet1/2>, <Interface: Ethernet1/48>, <Interface: Ethernet1/49>, <Interface: Ethernet1/50>, <Interface: Ethernet1/54>, <Interface: mgmt0>, <Interface: port-channel1>, <Interface: port-channel2>, <Interface: port-channel3>, <Interface: port-channel201>]>

and the query /ipam/ip-addresses/?device_id=123 will return a similar queryset to the following in nbshell

>>> IPAddress.objects.filter(interface__device_id=123 )
<RestrictedQuerySet [<IPAddress: 192.0.2.12/24>]>


When you are working on an IPAddress queryset you will get IPAddress objects and when you are working on an Interface queryset you will get Interface objects they don't have all the same fields but they do have a few, you could put if/then in your template to have different behavior, but you are only going to be seeing one object type at a time


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by reduzent** [reply]
OK, I think I understand now. Defining many content types in the template makes the template available under different paths. The content of `queryset` depends on the path I call (whether I call `/ipam/ip-addresses` or `/dcim/interfaces`). 

Many thanks for the detailed explanation.
